sign/eddilithium3/eddilithium.go

```diff
@@ -0,0 +1,210 @@
+// eddilithium3 implements the hybrid signature scheme Ed25519-Dilithium3.
+package eddilithium3
+
+import (
+	"crypto"
+	cryptoRand "crypto/rand"
+	"errors"
+	"io"
+
+	"github.com/cloudflare/circl/internal/shake"
+	"github.com/cloudflare/circl/sign/dilithium/mode3"
+	"github.com/cloudflare/circl/sign/ed25519"
+)
+
+const (
+	// Size of seed for NewKeyFromSeed
+	SeedSize = mode3.SeedSize // = ed25519.SeedSize = 32
+
+	// Size of a packed PublicKey
+	PublicKeySize = mode3.PublicKeySize + ed25519.PublicKeySize
+
+	// Size of a packed PrivateKey
+	PrivateKeySize = mode3.PrivateKeySize + ed25519.PrivateKeySize
+
+	// Size of a signature
+	SignatureSize = mode3.SignatureSize + ed25519.SignatureSize
+)
+
+// PublicKey is the type of an EdDilithium3 public key.
+type PublicKey struct {
+	e ed25519.PublicKey
+	d mode3.PublicKey
+}
+
+// PrivateKey is the type of an EdDilithium3 private key.
+type PrivateKey struct {
+	e ed25519.KeyPair
+	d mode3.PrivateKey
+}
+
+// GenerateKey generates a public/private key pair using entropy from rand.
+// If rand is nil, crypto/rand.Reader will be used.
+func GenerateKey(rand io.Reader) (*PublicKey, *PrivateKey, error) {
+	var seed [SeedSize]byte
+	if rand == nil {
+		rand = cryptoRand.Reader
+	}
+	_, err := io.ReadFull(rand, seed[:])
+	if err != nil {
+		return nil, nil, err
+	}
+
+	pk, sk := NewKeyFromSeed(&seed)
+	return pk, sk, nil
+}
+
+// NewKeyFromSeed derives a public/private key pair using the given seed.
+func NewKeyFromSeed(seed *[SeedSize]byte) (*PublicKey, *PrivateKey) {
+	var seed1 [32]byte
+	var seed2 [32]byte
+
+	h := shake.NewShake256()
+	_, _ = h.Write(seed[:])
+	_, _ = h.Read(seed1[:])
+	_, _ = h.Read(seed2[:])
+	dpk, dsk := mode3.NewKeyFromSeed(&seed1)
+	epair := ed25519.NewKeyFromSeed(seed2[:])
+
+	return &PublicKey{epair.GetPublic(), *dpk}, &PrivateKey{*epair, *dsk}
+}
+
+// SignTo signs the given message and writes the signature into signature.
+// It will panic if signature is not of length at least SignatureSize.
+func SignTo(sk *PrivateKey, msg []byte, signature []byte) {
+	mode3.SignTo(
+		&sk.d,
+		msg,
+		signature[:mode3.SignatureSize],
+	)
+	esig, _ := sk.e.SignPure(
+		msg,
+	)
+	copy(signature[mode3.SignatureSize:], esig[:])
+}
+
+// Verify checks whether the given signature by pk on msg is valid.
+func Verify(pk *PublicKey, msg []byte, signature []byte) bool {
+	if !mode3.Verify(
+		&pk.d,
+		msg,
+		signature[:mode3.SignatureSize],
+	) {
+		return false
+	}
+	if !ed25519.Verify(
+		pk.e,
+		msg,
+		signature[mode3.SignatureSize:],
+	) {
+		return false
+	}
+	return true
+}
+
+// Unpack unpacks pk to the public key encoded in buf.
```

Comment:
```suggestion
// Unpack pk to the public key encoded in buf.
```

Reply:
I prefer your suggestion, but the example from the code guidelines you linked is

```go
// Encode writes the JSON encoding of req to w.
func Encode(w io.Writer, req *Request) { ...
```

Wouldn't they have used `Encode req as JSON and write into w.` or something of that affect?

Reply:
Also your sentence is an imperative.  Wouldn't `Unpacks pk ...` make more sense?

Reply:
First word has to be `Unpack`, and it just doesn't read well to use `unpack` twice.

Reply:
In the other example, at least `write` is actually a different word :joy: 

Anyway, the suggested edit I made is what I do in this annoying situation. I'll leave it up to you now, I don't want to bother you endlessly with small comments.

Reply:
@armfazh What is your opinion?  Personally, I don't like to strictly adhere to this guideline strictly as it distract us with the occasional linguistic puzzle as above.

Reply:
I personally follow that guideline, the syntax is 

```
// XYZ <whatever> <PERIOD>
 function XYZ()
```

This works well for functions and variables, but the issue is when the name of the function is a verb, or when the function is a method as in object-oriented programming.

For `CIRCL`, I think it's enough to have a documentation string that is easy to read and talks about its input parameters and what it returns.

Reply:
Feel free to open an issue for discussing style and documentation, if needed.